cli/util/sync/async_flag.rs

```diff
@@ -15,14 +15,10 @@ impl Default for AsyncFlag {
 
 impl AsyncFlag {
   pub fn raise(&self) {
-    self.0.close();
-  }
-
-  pub fn is_raised(&self) -> bool {
-    self.0.is_closed()
+    self.0.add_permits(1);
   }
 
   pub async fn wait_raised(&self) {
-    self.0.acquire().await.unwrap_err();
+    drop(self.0.acquire().await);
   }
 }
```

Comment:
`init_flag` handling was wrong. New requests queued just after it was raised would skip the queue of unblocked tasks. So we now preserve the semaphore throughout standard operation.

Reply:
It was especially visible with notebooks due to the timing but this probably fixes some other flakes e.g. I'll take another look at #21770 later.